controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller_intg_test.go

```diff
@@ -0,0 +1,230 @@
+						"appname": "db",
+					},
+				},
+				Template: vmopv1.VirtualMachineTemplateSpec{
+					ObjectMeta: vmopv1common.ObjectMeta{
+						Labels: map[string]string{
+							"appname": "db",
+						},
+						Annotations: make(map[string]string),
+					},
+					Spec: vmopv1.VirtualMachineSpec{
+						ImageName:  "dummy-image",
+						ClassName:  "dummy-class",
+						PowerState: vmopv1.VirtualMachinePowerStateOn,
+						Network: &vmopv1.VirtualMachineNetworkSpec{
+							Interfaces: []vmopv1.VirtualMachineNetworkInterfaceSpec{
+								{
+									Name: "eth0",
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		rsKey = types.NamespacedName{Name: rs.Name, Namespace: rs.Namespace}
+	})
+
+	AfterEach(func() {
+		ctx.AfterEach()
+		ctx = nil
+		intgFakeVMProvider.Reset()
+	})
+
+	getVirtualMachineReplicaSet := func(ctx *builder.IntegrationTestContext, objKey types.NamespacedName) *vmopv1.VirtualMachineReplicaSet {
+		rs := &vmopv1.VirtualMachineReplicaSet{}
+		if err := ctx.Client.Get(ctx, objKey, rs); err != nil {
+			return nil
+		}
+		return rs
+	}
+
+	waitForReplicaSetFinalizer := func(ctx *builder.IntegrationTestContext, objKey types.NamespacedName) {
+		EventuallyWithOffset(1, func() []string {
+			if rs := getVirtualMachineReplicaSet(ctx, objKey); rs != nil {
+				return rs.GetFinalizers()
+			}
+			return nil
+		}).Should(ContainElement(finalizerName), "waiting for VirtualMachineReplicaSet finalizer")
+	}
+
+	ensureReplicas := func(ctx *builder.IntegrationTestContext, labels map[string]string, desiredReplicas int) {
+		Eventually(func(g Gomega) int {
+			vmList := &vmopv1.VirtualMachineList{}
+			err := ctx.Client.List(ctx, vmList, client.InNamespace(ctx.Namespace), client.MatchingLabels(labels))
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(vmList).ToNot(BeNil())
+			return len(vmList.Items)
+		}, 10*time.Second, 1*time.Second).Should(Equal(desiredReplicas))
+	}
+
+	Context("Reconcile", func() {
+		dummyInstanceUUID := "instanceUUID1234"
+
+		BeforeEach(func() {
+			intgFakeVMProvider.Lock()
+			intgFakeVMProvider.CreateOrUpdateVirtualMachineFn = func(ctx context.Context, vm *vmopv1.VirtualMachine) error {
+				// Used below just to check for something in the Status is updated.
+				vm.Status.InstanceUUID = dummyInstanceUUID
+				return nil
+			}
+			intgFakeVMProvider.Unlock()
+		})
+
+		AfterEach(func() {
+			By("Delete VirtualMachineReplicaSet", func() {
+				if err := ctx.Client.Delete(ctx, rs); err == nil {
+					rs := &vmopv1.VirtualMachineReplicaSet{}
+					// If ReplicaSet is still around because of finalizer, try to cleanup for next test.
+					if err := ctx.Client.Get(ctx, rsKey, rs); err == nil && len(rs.Finalizers) > 0 {
+						rs.Finalizers = nil
+						_ = ctx.Client.Update(ctx, rs)
+					}
+				} else {
+					Expect(apierrors.IsNotFound(err)).To(BeTrue())
+				}
+			})
+
+			// Fake client doesn't support garbage collection, so we can't verify that all replicas have been cleaned up.
```

Comment:
The fake client is not used for integration tests. It is because env-test only runs apiserver and etcd and GC is done by kube-controller-manager

Reply:
TIL.  Will update the comment to reflect this.